src/java.base/share/classes/java/util/concurrent/Executors.java

```diff
@@ -824,14 +823,25 @@ public <T> T invokeAny(Collection<? extends Callable<T>> tasks,
         }
     }
 
-    private static class FinalizableDelegatedExecutorService
+    /**
+     * A DelegatedExecutorService that uses a Cleaner to shut down the underlying
+     * ExecutorService when the wrapper becomes phantom reachable.
+     */
+    private static class AutoShutdownDelegatedExecutorService
             extends DelegatedExecutorService {
-        FinalizableDelegatedExecutorService(ExecutorService executor) {
+        private final Cleanable cleaner;
+        AutoShutdownDelegatedExecutorService(ExecutorService executor) {
             super(executor);
+            Runnable action = () -> {
+                PrivilegedAction<Void> pa = () -> { executor.shutdown(); return null; };
+                @SuppressWarnings("removal")
+                var ignore = AccessController.doPrivileged(pa);
+            };
+            cleaner = CleanerFactory.cleaner().register(this, action);
         }
-        @SuppressWarnings("removal")
-        protected void finalize() {
-            super.shutdown();
+        @Override
+        public void shutdown() {
+            cleaner.clean();
```

Comment:
Hmmm... so now shutdown no longer requires permission. You should probably call super.shutdown() before invoking cleaner.clean(), assuming that shutdown() can be called twice.

Reply:
> Hmmm... so now shutdown no longer requires permission. You should probably call super.shutdown() before invoking cleaner.clean(), assuming that shutdown() can be called twice. You could modify the cleanup action to only call shutdown() if isShutdown() returns false.

That's right. It will be a non-issue once the SM goes away but while execution mode is supported then the shutdown method and the cleaner action have to be different (as it was initially).

Reply:
The cleaning action would not have access to the isShutdown() instance method of the (Phantom-reachable) AutoShutdownDelegatedExecutorService.

Reply:
> The cleaning action would not have access to the isShutdown() instance method of the (Phantom-reachable) AutoShutdownDelegatedExecutorService.

The cleaning action has a reference to the delegate (the underlying ExecutorService) so it can test if it shutdown as Daniel's suggests - it's more of an optimization to avoid doing a second call to shutdown in a privileged action.

Reply:
OK, right